### PR TITLE
** noaspect **

### PR DESCRIPTION
**Pull request by GingerAdams**
Added / exposed ** noaspect ** parameter to allow cusom aspect ratios.
e.g. 16:9 over PAL Composite output for previews and overlays.

### CODE REVIEW

**Comment by mrotter12** [comment]
Hello,
I am quite new using the picamera module.
Thank you very much for the "noaspect" addition !
Works great !
But I dont get it to work with an overlay image.
Is it possible to post an example code ?

That would be great :-)
Thank you in advance,
Mike

**Comment by waveform80** [comment]
@GingerAdams intriguing! And very timely, given I was just working through the picamera tickets and this seems to address #440! A few requests on the PR:

* No need to change the copyright dates :)
* No need to add anything to the PiCamera class; the `preview_window`, `preview_fullscreen` and other attributes exist solely for backwards compatibility (the [Deprecated Functionality](http://picamera.readthedocs.io/en/release-1.13/deprecated.html) chapter spells out the gory details).

In other words, the only changes needed are the changes you've made in the `renderers` module. Incidentally, those look great: you've got the getter, setter and property decl exactly as I would've done them. My one and only concern is the name of the property: noaspect. I realize you're basically copying the MMAL layer in the naming but generally I try and avoid "negative" properties because they're (slightly) trickier for people to reason about. For example, having an "enabled" property (True is on, False is off) is generally easier for people than a "not_enabled" property (where someone then has to consider "False means it's on"). In this case, I'd probably name the property "anamorphic". Admittedly it's an obscure term, but it's probably one that, if people aren't familiar with it, they ought to learn it (in the context of messing around with a video camera), and we can always add a link in the docs to the relevant [wikipedia article](https://en.wikipedia.org/wiki/Anamorphic_format) (incidentally, thank you *very* much for including docs!).

If you're up to adjusting the PR yourself (which can simply be done by adding more commits to your branch), please do so. However, GitHub has marked this "First-time GitHub contributor" so I'd certainly understand if you're not necessarily comfortable with modifying a PR; in that case I'm happy to merge this as is and tweak bits myself (although it'll probably be tomorrow before I get a chance to play with it myself).

Either way, thanks for the PR!

**Comment by GingerAdams** [comment]
Dave,

Thanks for getting in touch.

Agreed, was never too comfortable with the term myself for the same reasons you have mentioned and you are correct, I chose it simply because that is what it was called. "anamorphic" seems appropriate and less ambiguous.

As for the docs. Didn't realise I'd done anything special. If someone's taken the time to show you how the like it done it seems only right that you respect that and continue in the same manor, especially if you take from the open source community more than you give like me.

Yes I'm a github virgin, this is the first open source thing I've ever delved into. It's been a while since I raised the PR and as such may take me a little while to close as I've never done it before. I'd also need to blow out a few cobwebs as to what I did. I'm sure it's quite straightforward just never done it. Feel free to expedite it if you need to.


On 9 Sep 2017 11:20 pm, Dave Jones <notifications@github.com> wrote:

@GingerAdams<https://github.com/gingeradams> intriguing! And very timely, given I was just working through the picamera tickets and this seems to address #440<https://github.com/waveform80/picamera/issues/440>! A few requests on the PR:

  *   No need to change the copyright dates :)
  *   No need to add anything to the PiCamera class; the preview_window, preview_fullscreen and other attributes exist solely for backwards compatibility (the Deprecated Functionality<http://picamera.readthedocs.io/en/release-1.13/deprecated.html> chapter spells out the gory details).

In other words, the only changes needed are the changes you've made in the renderers module. Incidentally, those look great: you've got the getter, setter and property decl exactly as I would've done them. My one and only concern is the name of the property: noaspect. I realize you're basically copying the MMAL layer in the naming but generally I try and avoid "negative" properties because they're (slightly) trickier for people to reason about. For example, having an "enabled" property (True is on, False is off) is generally easier for people than a "not_enabled" property (where someone then has to consider "False means it's on"). In this case, I'd probably name the property "anamorphic". Admittedly it's an obscure term, but it's probably one that, if people aren't familiar with it, they ought to learn it (in the context of messing around with a video camera), and we can always add a link in the docs to the relevant wikipedia article<https://en.wikipedia.org/wiki/Anamorphic_format> (incidentally, thank you very much for including docs!).

If you're up to adjusting the PR yourself (which can simply be done by adding more commits to your branch), please do so. However, GitHub has marked this "First-time GitHub contributor" so I'd certainly understand if you're not necessarily comfortable with modifying a PR; in that case I'm happy to merge this as is and tweak bits myself (although it'll probably be tomorrow before I get a chance to play with it myself).

Either way, thanks for the PR!

**Comment by GingerAdams** [comment]
I'll see what I can dig out. And let you know.

On 9 Sep 2017 4:06 pm, mrotter12 <notifications@github.com> wrote:

Hello,
I am quite new using the picamera module.
Thank you very much for the "noaspect" addition !
Works great !
But I dont get it to work with an overlay image.
Is it possible to post an example code ?

That would be great :-)
Thank you in advance,
Mike

**Comment by GingerAdams** [comment]
@mrotter12 
Try this (see below). I've hacked it from some development I was doing but is currently on the back burner.
Be aware you may need to play around with it as I haven't had a chance to test it out of context and the code itself was ad-hoc and unrefined so you may come across some anomalies. Also be aware, the correlation between Camera resolution / position, preview resolution / position, overlay resolution / position and cropping are tentative and not something I fully got to grips with, so again, play with it a bit to extend your understanding and make sure you read the PiCamera docs.

Good Luck

CODE:
import picamera
import numpy as np
from PIL import Image, ImageDraw, ImageFont, ImageColor

# Screen
resX, resY = 720, 540 # screen resolution for preview 4:3 aspect

# Overlay
overResX, overResY = 768, 446 # Overlay Resolution x // 32 and y // 16 with 4:3 aspect ratio

# PiCamera
camera = picamera.PiCamera() # Define PiCamera object
camera.resolution = (1296, 972)#(1296, 972) #FoV sensor area (2596x1944 native, 1296x972 binned, 640x480 binned) 4:3 aspect
camera.brightness = 50 # Set Brightness
camera.contrast = 0 # Set Contrast
camera.framerate = 40 # Set required framerate for video
camera.led = False
camera.rotation = 180
# Camera modes
camera.exposure_mode = 'auto'

# Start Preview
camera.start_preview(fullscreen=False,
                     window=(0, 20, resX, resY),
                     noaspect=True,
                     crop=(0, 126, 1296, 730))# enable preview to primary display with 16:9 crop

# Create an image for the overlay
# Colours
blackt = '#00000000' # Transparency colour
foreColour = 'white'

# Image Canvas
img = Image.new('RGBA', (overResX, overResY), blackt)
draw = ImageDraw.Draw(img)

#Fill entire display to clear
draw.rectangle([0, 0, overResX, overResY],
               blackt, foreColour)
# Add Roi frame
draw.rectangle([36, 20, overResX - 15, overResY - 16],
               blackt, foreColour)

# Overlay(s)
olStatic = camera.add_overlay(img.tobytes(),
                              format='rgba',
                              size=(overResX, overResY),
                              layer=3,
                              alpha=255,
                              noaspect=True)

#---------------------------------------------------------------------------------------------------

try:
    while True:
        i += 1 # Just loop
        
finally:
    camera.remove_overlay(olStatic)
    camera.close()

**Comment by atomiclama** [comment]
Great thanks this is exactly what I need.
@waveform80 Are there any timescales to get this commited to master?

In the meantime, in order to use it can I just replace the .py files in /usr/share/pyshared/picamera with those from this PR?

**Comment by GingerAdams** [comment]
Hi Andy,

No timescale. Can do it whenever just not sure how, never done it before.

Yep, just copy them over. There's nothing pending.

Enjoy.

Regards

Stuart


On 13 Dec 2017 3:13 pm, Andy <notifications@github.com> wrote:

Great thanks this is exactly what I need.
@waveform80<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fwaveform80&data=02%7C01%7C%7C7feb998d839d427523ee08d5423c0d50%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636487748036732898&sdata=0pjBlg7btB2OTne8FJWgoC6Z4LA2LaqKDf3QYckVmVc%3D&reserved=0> Are there any timescales to get this commited to master?

In the meantime, in order to use it can I just replace the .py files in /usr/share/pyshared/picamera with those from this PR?

**Comment by atomiclama** [comment]
New is to python so not sure where to start looking. I have checked out GingerAdams:master and copied all the .py files to "/usr/lib/python2.7/dist-packages/picamera/" now I get this error, any ideas where to start looking.

Traceback (most recent call last):
  File "./camera.py", line 65, in <module>
    camera.start_preview()
  File "/usr/lib/python2.7/dist-packages/picamera/camera.py", line 806, in start_preview
    self, self._camera.outputs[self.CAMERA_PREVIEW_PORT], **options)
TypeError: __init__() got an unexpected keyword argument 'noaspect'

**Comment by atomiclama** [comment]
@GingerAdams as far as I'm aware you can't perform the merges that is down to waveform80 the repo owner.

In the mean time I'll try to get it working on my setup.

**Comment by atomiclama** [comment]
@GingerAdams I think you also need to change in camera.py noaspect -> anamorphic

**Comment by atomiclama** [comment]
@GingerAdams HI fella, struggling here. When I set 
camera.start_preview(anamorphic=True)

It makes no difference to the display output. I'm using the composite PAL output through to a small monitor. 
Is there anything you can suggest that I try ?

**Comment by GingerAdams** [comment]
Mmm. I think you may need to be in windowed mode. Sorry to be vague, It's been a while since I've played with it.

On 14 Dec 2017 3:06 pm, Andy <notifications@github.com> wrote:

@GingerAdams<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgingeradams&data=02%7C01%7C%7C18174783cc68424c0ded08d543044548%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636488607967966336&sdata=LPa7kVaTM9aEA2tlqdDW4xkhCf8Fo7t7Vua4zHWl3xs%3D&reserved=0> HI fella, struggling here. When I set
camera.start_preview(anamorphic=True)

It makes no difference to the display output. I'm using the composite PAL output through to a small monitor.
Is there anything you can suggest that I try ?

**Comment by atomiclama** [comment]
Ok the problem lies with the refactor from noaspect to anamorphic.

I believe that the section of renderers.py still needs to be noaspect as that is the variable in the MMAL side.
`def _get_anamorphic(self):
        return self.renderer.inputs[0].params[mmal.MMAL_PARAMETER_DISPLAYREGION].noaspect.value != mmal.MMAL_FALSE
    def _set_anamorphic(self, value):
        mp = self.renderer.inputs[0].params[mmal.MMAL_PARAMETER_DISPLAYREGION]
        mp.set = mmal.MMAL_DISPLAY_SET_NOASPECT
        mp.noaspect = bool(value)
        self.renderer.inputs[0].params[mmal.MMAL_PARAMETER_DISPLAYREGION] = mp`

File: picamera/renderers.py [hunk 1]
```diff
@@ -177,6 +178,23 @@ def _set_fullscreen(self, value):
         are active.
         """)
 
+    def _get_anamorphic(self):
+        return self.renderer.inputs[0].params[mmal.MMAL_PARAMETER_DISPLAYREGION].anamorphic.value != mmal.MMAL_FALSE
+    def _set_anamorphic(self, value):
+        mp = self.renderer.inputs[0].params[mmal.MMAL_PARAMETER_DISPLAYREGION]
+        mp.set = mmal.MMAL_DISPLAY_SET_NOASPECT
+        mp.anamorphic = bool(value)
```
[the remark this answers is not in the document]
Comment (GingerAdams) [reply picamera/renderers.py]: Done.

Thanks for the spot.

**Comment by waveform80** [comment]
Wow - sorry this has been so long - I actually remember replying to this initially. Anyway, definitely worth inclusion in 1.14.

**Comment by GingerAdams** [comment]
Thanks. Good to know I've finally given a little something back.
________________________________

**Pull request merged**